File: java/easy/338_counting_bits.java
// https://leetcode.com/problems/counting-bits/description/
package java.easy;

class CountingBits {
    public int[] countBits(int n) {
        int[] result = new int[n+1];
        result[0] = 0;

        for (int i = 1; i <= n; i++) {
            int sum = 0;
            int tmp = i;
            while (tmp > 0) {
                sum++;
                tmp &= (tmp-1);
            }
            result[i] = sum;
        }

        return result;
    }
}

File: go/easy/338_counting_bits.go
// https://leetcode.com/problems/counting-bits/
package easy

func CountBits(n int) []int {
	result := make([]int, n+1)
	for i := 0; i <= n; i++ {
		result[i] = popCount(i)
	}

	return result
}

func popCount(idx int) int {
	count := 0
	for count = 0; idx != 0; count++ {
		idx = idx & (idx - 1)
	}

	return count
}

// DP approach
// func CountBits(n int) []int {
//     dp := make([]int, n+1)
//     for i := 0; i < n+1; i++ {
//         dp[i] = dp[i/2] + (i%2)
//     }
//     return dp
// }
